src/builtins_str.rs

```diff
@@ -787,6 +789,104 @@ fn builtin_char_is_whitespace(
     ))
 }
 
+fn builtin_char_int(
+    environment: &mut Environment,
+    args: &mut dyn Iterator<Item = Expression>,
+) -> Result<Expression, LispError> {
+    let next_arg = param_eval(environment, args, "char->int")?;
+    let next_arg_d = next_arg.get();
+    params_done(args, "char->int")?;
+    let to_int = |s: &str| -> Result<Expression, LispError> {
+        let mut count = 0;
+        for _s in <str as UnicodeSegmentation>::graphemes(s, true) {
+            count += 1;
+        }
+        match count {
+            0 => Ok(Expression::alloc_data(ExpEnum::Int(0))),
+            1 => {
+                let mut int_val: u32 = 0;
+                let mut overflow = false;
+                for c in s.chars() {
+                    let (add, overflowed) = int_val.overflowing_add(c as u32);
+                    if overflowed {
+                        overflow = true;
+                        break;
+                    }
+                    int_val += add;
+                }
+                if overflow {
+                    Err(LispError::new("error, overflow occurred adding unicode scalar values that compose provided grapheme interpreted as unsigned 32 byte integers: {:?}."))
+                } else {
+                    Ok(Expression::alloc_data(ExpEnum::Int(int_val as i64)))
+                }
+            }
+            _ => Err(LispError::new(
+                "function takes one grapheme, multiple were provided.",
+            )),
+        }
+    };
+    match &next_arg_d.data {
+        ExpEnum::String(s, _) => to_int(s),
+        ExpEnum::Char(s) => to_int(s),
+        _ => Err(LispError::new(
+            "expected one argument of type Char or String",
+        )),
+    }
+}
+
+fn builtin_unicode_scalars(
+    environment: &mut Environment,
+    args: &mut dyn Iterator<Item = Expression>,
+) -> Result<Expression, LispError> {
+    let next_arg = param_eval(environment, args, "unicode-scalars")?;
+    let next_arg_d = next_arg.get();
+    params_done(args, "unicode-scalars")?;
+    let with_str = |s: &str| {
+        let mut unicode_scalars: Vec<String> = Vec::new();
+        for c in s.chars() {
+            unicode_scalars.push(format!("{}", c.escape_unicode()));
+        }
+
+        let mut strings = Vec::with_capacity(unicode_scalars.len());
+        for s in unicode_scalars {
+            strings.push(Expression::alloc_data(ExpEnum::String(
+                s.to_string().into(),
+                None,
+            )));
+        }
+        Ok(Expression::alloc_data(ExpEnum::Vector(strings)))
+    };
+    match &next_arg_d.data {
+        ExpEnum::String(s, _) => with_str(s),
+        ExpEnum::Char(s) => with_str(s),
+        _ => Err(LispError::new(
+            "expected one argument of type Char or String",
+        )),
+    }
+}
+
+fn builtin_str_hash(
+    environment: &mut Environment,
+    args: &mut dyn Iterator<Item = Expression>,
+) -> Result<Expression, LispError> {
```

Comment:
Should this be str->float instead of str-hash?  Hashes are typically ints of some sort (I think) and we do not expose any hashing stuff in general (might be misleading). I like the f64 solution to the u/i64 issue.

Reply:
Maybe str-hash is fine, we can talk about it.